### PR TITLE
Default to desktop when display plugins fail to load

### PR DESCRIPTION
**Pull request by zzmp**
- Add logging for oculus SDK errors.
- Check plugins when they are activated.
- Add fallback logic for display plugin changes.

### CODE REVIEW

**Comment by zzmp** [comment]
@cozza13 I'm posting this up for dev review, but I want to test the code myself before I put up a `QA Ready` label.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -4892,7 +4892,17 @@ void Application::updateDisplayMode() {
 
         // FIXME probably excessive and useless context switching
         _offscreenContext->makeCurrent();
-        newDisplayPlugin->activate();
+
+        // If the new plugin fails to activate, fallback to first item on the list
+        if (!newDisplayPlugin->activate()) {
```
Comment (jherico) [inline interface/src/Application.cpp]: if `newDisplayPlugin == displayPlugins.at(0);` then you've entered an infinite loop here.

File: plugins/hifiSdl2/src/SDL2Manager.h [hunk 2]
```diff
@@ -34,9 +34,7 @@ class SDL2Manager : public InputPlugin {
     virtual void init() override;
     virtual void deinit() override;
 
-    /// Called when a plugin is being activated for use.  May be called multiple times.
-    virtual void activate() override;
-    /// Called when a plugin is no longer being used.  May be called multiple times.
+    virtual bool activate() override;
```
Comment (jherico) [inline plugins/hifiSdl2/src/SDL2Manager.h]: please restore the comments.

Comment (zzmp) [inline plugins/hifiSdl2/src/SDL2Manager.h]: These are overrides, and the comments are on the base class. Still want them back?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 37a725b06dc8b9bc183609a2077ec54602c939bf available at: https://gist.github.com/8074627bd50839362db9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.dmg

File: plugins/hifiSdl2/src/SDL2Manager.h (hunk 2 above)
Comment (ZappoMan) [inline plugins/hifiSdl2/src/SDL2Manager.h]: yes please

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -4892,7 +4892,17 @@ void Application::updateDisplayMode() {
...
+            qWarning() << "Failed to activate plugin: " << newDisplayPlugin->getName();
+            newDisplayPlugin = displayPlugins.at(0);
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: wouldn't it be better to go back to the LAST active plugin?

Comment (zzmp) [inline interface/src/Application.cpp]: @ZappoMan My thinking was that desktop was the safest bet, and the most likely to always work.
It also does the least mouse capture, so would make it easiest to switch again.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b94e244844c99a8acf402634a31c0d8a62b115cc available at: https://gist.github.com/8074627bd50839362db9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.dmg

File: interface/src/Application.cpp (hunk 3 above)
Comment (ZappoMan) [inline interface/src/Application.cpp]: ok - I am convinced.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 34a77b47a34a2f59d9fff4f0e4a6b78a0c3f4750 available at: https://gist.github.com/8074627bd50839362db9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.dmg

**Comment by zzmp** [comment]
#### Intent

If a display plugin fails to load, the presentation thread will crash the system.
This fixes it so that instead of overall failure and log flooding, we fallback to the last display (if available), and then to the desktop display.

#### Testing

Do something to make Oculus (or OpenVR, I think that's Vive) fail *when switching to it*. This **does not** affect display failures when it is already active, only when it is first switched to. You should test two cases:
- Starting in HMD (with failure), and see it go to Desktop.
- Switching to HMD from something else (like 3D TV), and see it go back to whatever you had.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7bcef1d3194cbfd7e7dc995b067df3d2db7a1df7 available at: https://gist.github.com/8074627bd50839362db9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7493.dmg

**Comment by cozza13** [comment]
tested. Crashed the Oculus (powered off while interface was running). The display did not let me switch to Rift. In production is lets you do this and it will freeze the display. QA Pass

**Pull request merged**